### ISSUE TITLE
Deprecate `appData` in v4

### ISSUE THREAD

**Issue by piranna** [versatica/mediasoup#821]
## Feature Request

For upcoming v4, deprecate and remove `appData` field altogether. The same way there's no lists of elements on purposse since people can listen the `observer` objects and control the objects that has been created, the same can be done with `appData`, keeping in application logic a map of the elements and their associated info.

Alternatively, if we don't want to fully remove it (or as a transition step), enforce that all the info stored in `appData` is JSON-serializable, or at least that can be cloned with the `structuredCopy` algorythm. This would prevent to have "live" and untransferable objects stored inside `appData`, enforcing users to store instead some IDs and reference them with the actual objects in their application logic.

Also, this would lead code to be more strictly typed instead of using an `any` type (that infact, being a non-nullable object, it should be properly declared as `Record<string, unknown>`, by the way), leading to fewer bugs and errors for everybody.

**Comment by ibc** [comment]
appData is the only way to match calls to sendTransport.produce() with the transport.on("produce") event in mediasoup-client, so it has a value. In addition, we use it extensively in server side to attach additional custom info into producers, consumers, transports, etc. Whether users store jsonizable data into them or not is up to them, we cannot or should not enforce it because there is no reason to enforce it (it's app custom data). I don't see any actionable thing here, appData has a purpose and I see no reason to remove it or restrict it.

**Comment by ibc** [comment]
Regarding TS and Record<string, unknown>, I fail to see what the difference with declaring it an object (as it is now) is. Let's discuss that in case it can be changed, but appData won't be removed so let's close this.

**Comment by piranna** [comment]
> Regarding TS and Record<string, unknown>, I fail to see what the difference with declaring it an object (as it is now) is.

I get warnings from Typescript when declaring something like `Object`, because it says is too generic and it's like "accept anything not nullable". This warning makes sense, since we could set it to any object, while in code is defined with `{}`. It recommends to use `Record<string, unknown>` since explicitly declares a mapping object with string keys and unknown values, that later will be type enforced when accessed, instead of "any object". I agree with it that's more type safe than using `any` or `Object` as type here.

TL;DR: we want to ensure it's defined and used as a plain Javascript object literal (`{}`), not any Javascript object (Map, Set, Date, Worker, HTMLElement...), and it doesn't matter what we set as values of this object literal, but we want to ensure they are typecheck later. Using `Record<string, unknown>` fix both things, while using `Object` or `any` fix any of them.

**Comment by ibc** [comment]
appData is readonly and initialized to {} so I fail to see why TS would complain. Which version of TS are you using?

**Comment by ibc** [comment]
forget my last comment. Ongoing PR.

**Comment by ibc** [comment]
Actually I pushed directly to `v3` brach: https://github.com/versatica/mediasoup/commit/0c07ef60d4f2011f7afe84b357ab778dd829c908

**Comment by piranna** [comment]
> Actually I pushed directly to `v3` brach: [0c07ef6](https://github.com/versatica/mediasoup/commit/0c07ef60d4f2011f7afe84b357ab778dd829c908)

Oh, great, it looks really nice :-D
